# 🛒 Walmart Sales Data Analysis  

# =========================================
# Author: Babatunde Dosunmu
# =========================================

# This project analyzes Walmart sales data to uncover:
- Sales trends over time
- Customer purchasing behavior
- Employee performance
- Payment method insights

 *Goal: Practice data cleaning, visualization, and descriptive analytics using Python.*

In [6]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [ ]:
# Load dataset
df = pd.read_csv(r"C:\Users\DELL\OneDrive\Documents\my_project\walmart\walmart\Walmart.csv")
df.head()

sale_id sale_reference  product_id  customer_id  employee_id  quantity  \
0        1       5ea72efb          55          499            1         1   
1        2       b5015710          19          212           15         3   
2        3       f85add14         145          273           14         4   
3        4       0de53585          81          258            5         1   
4        5       0471cfb9          38           52           14         5   

          sale_date          state                  town method_of_payment  
0   12/31/2022 3:43        Indiana             Aprilstad       Credit Card  
1   3/14/2022 15:49       Missouri  West Angelicaborough        Debit Card  
2   12/17/2020 7:57  New Hampshire     New Dominiqueland       Credit Card  
3    2/28/2024 9:21      Tennessee             Erikburgh              Cash  
4  10/15/2023 18:53       Colorado           Johnsonfort    Mobile Payment

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   sale_id            20000 non-null  int64 
 1   sale_reference     20000 non-null  object
 2   product_id         20000 non-null  int64 
 3   customer_id        20000 non-null  int64 
 4   employee_id        20000 non-null  int64 
 5   quantity           20000 non-null  int64 
 6   sale_date          20000 non-null  object
 7   state              20000 non-null  object
 8   town               20000 non-null  object
 9   method_of_payment  20000 non-null  object
dtypes: int64(5), object(5)
memory usage: 1.5+ MB


#### Key Points
- No missing values (all columns have 20,000 non-null entries).
- 5 columns are numeric (`int64`), 5 columns are strings (`object`)
- `sale_date` should be converted to **datetime** for analysis. 

In [11]:
df.describe()

sale_id    product_id  customer_id   employee_id      quantity
count  20000.000000  20000.000000  20000.00000  20000.000000  20000.000000
mean   10000.500000     75.593900    251.17830      7.994050      3.003850
std     5773.647028     43.105583    144.97009      4.324506      1.418145
min        1.000000      1.000000      1.00000      1.000000      1.000000
25%     5000.750000     39.000000    125.00000      4.000000      2.000000
50%    10000.500000     76.000000    253.00000      8.000000      3.000000
75%    15000.250000    112.000000    375.00000     12.000000      4.000000
max    20000.000000    150.000000    500.00000     15.000000      5.000000

In [ ]:
#### Key Point`sale_date` is converted from string to datetime for easier time-based analysis.
# Convert sale_date to datetime
df['sale_date'] = pd.to_datetime(df['sale_date'])

In [15]:
#time features
df['year'] = df['sale_date'].dt.year
df['month'] = df['sale_date'].dt.month
df['week'] = df['sale_date'].dt.isocalendar().week
df['weekday'] = df['sale_date'].dt.day_name()

In [17]:
# Revenue (fallback if no price column)
if "price" in df.columns:
    df['revenue'] = df['quantity'] * df['price']
else:
    df['revenue'] = df['quantity']

In [18]:
#Sales Trends
monthly_sales = df.groupby(['year','month'])['revenue'].sum().reset_index()

fig = px.line(monthly_sales, x="month", y="revenue", color="year",
              title="📈 Monthly Sales Revenue Trend")
fig.show()


## 1️⃣ Sales Trends  

### Total Sales Over Time  
- **What the plot shows:** Revenue across days/weeks/months.  
- **Interpretation:** 2025 (Blue) appears more stable and higher in some months

In [19]:
#costumer behavior
cust_trend = df.groupby(['year','month'])['customer_id'].nunique().reset_index()

fig = px.line(cust_trend, x="month", y="customer_id", color="year",
              title="👥 Active Customers Over Time")
fig.show()

if "method_of_payment" in df.columns:
    fig = px.pie(df, names="method_of_payment", title="💳 Payment Methods Distribution")
    fig.show()


## 2️⃣ Customer Insights  

### Active Customers Over Time  
- **What the plot shows:** How many unique customers shop daily/weekly.  
- **Interpretation:** A steady rise in unique customers means customer base is growing. If it fluctuates, it may point to seasonal shopping or inconsistent engagement.  

### Payment Methods Distribution  
- **What the plot shows:** Cash, card, and digital payment preferences.  
- **Interpretation:** If card/digital dominates, it suggests customers prefer convenience. If cash is high, there may be room for improving digital adoption.  

In [20]:
#Employee Performance
employee_sales = df.groupby('employee_id')['revenue'].sum().nlargest(10).reset_index()

fig = px.bar(employee_sales, x="employee_id", y="revenue",
             title="🏆 Top 10 Employees by Sales Revenue",
             text_auto=True)
fig.show()


### Top Products by Sales  
- **What the plot shows:** The most purchased products.  
- **Interpretation:** Some products sell a lot (fast-moving, lower price) while others bring in high revenue even if sold in smaller quantities (premium products). This helps identify *cash cow items vs. volume drivers*.  

In [ ]:
#KPIs
kpis = {
    "Total Revenue": round(df['revenue'].sum(), 2),
    "Average Quantity per Sale": round(df['quantity'].mean(), 2),
    "Unique Customers": df['customer_id'].nunique(),
    "Total Transactions": df['sale_id'].nunique()
}

{'Total Revenue': np.int64(60077),
 'Average Quantity per Sale': np.float64(3.0),
 'Unique Customers': 500,
 'Total Transactions': 20000}

# Key Insights from Walmart Sales Analysis

✅ **Sales Trends:** Revenue grows steadily with peak months clearly visible.  
✅ **Customer Insights:** Unique active customers are increasing monthly, with [insert dominant payment method].  
✅ **Employee Performance:** Employee [X] leads in total sales contribution.  
✅ **KPIs:**  

- **Total Revenue:** 60,077  
- **Average Quantity per Sale:** 3 items  
- **Unique Customers:** 500  
- **Total Transactions:** 20,000  

---
